### ISSUE TITLE
Rubygems / Bundler updates are moving dependencies from global to secondary source

### ISSUE THREAD

**Issue by agrobbin** [dependabot/dependabot-core#9326]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

Bundler

### Package manager version

Bundler 2.5.4

### Language version

Ruby 3.3.3

### Manifest location and content before the Dependabot update

* `Gemfile`
* `Gemfile.lock`

### dependabot.yml content

```yaml
version: 2

registries:
  bundler-github:
    type: rubygems-server
    url: https://rubygems.pkg.github.com/[redacted]
    token: [redacted]

  bundler-graphql-pro:
    type: rubygems-server
    url: https://gems.graphql.pro/
    token: [redacted]

updates:
  - package-ecosystem: 'bundler'
    registries:
      - bundler-github
      - bundler-graphql-pro
    vendor: true
    directory: '/'
    schedule:
      interval: 'daily'
      time: '10:00'
      timezone: 'America/New_York'
    versioning-strategy: 'increase-if-necessary'
    groups:
      google:
        patterns:
          - 'google*'
      graphql:
        patterns:
          - 'graphql*'
      packwerk:
        patterns:
          - 'packs*'
          - 'packwerk*'
      pry:
        patterns:
          - 'pry*'
      rubocop:
        patterns:
          - 'rubocop*'
      sentry:
        patterns:
          - 'sentry*'

  - package-ecosystem: 'github-actions'
    directory: '/'
    schedule:
      interval: 'daily'
      time: '10:00'
      timezone: 'America/New_York'
```

### Updated dependency

We've seen this with a couple of dependencies, but the most recent culprit is `shoulda-matchers` going from v6.1.0 to v6.2.0.

### What you expected to see, versus what you actually saw

Dependabot is moving a bunch of dependencies from our primary source (`https://rubygems.org/`) within `Gemfile.lock` to a secondary source, when there is an overlap of sub-dependencies between sources. That secondary source includes a gem that has the same sub-dependencies as the dependencies that were moved, which I *think* is the likely cause.

### Native package manager behavior

Bundler will, AFAIK, only list dependencies in a secondary source that are only available within that source. Sub-dependencies that are available in the global source are only listed in the global source within `Gemfile.lock`.

### Images of the diff or a link to the PR, issue, or logs

![github com_huntclub_atlas-rails_pull_1614_files](https://github.com/dependabot/dependabot-core/assets/46724/c0a61da2-8b66-4d11-b1a4-f1734ddf524d)

I hope this diff screenshot is helpful! The removals are within the `https://rubygems.org/` source, while the additions are within a private source (in our case, a GItHub Packages RubyGems registry).

### Smallest manifest that reproduces the issue

```ruby
source 'https://rubygems.pkg.github.com/[redacted]' do
  gem 'foo' # <= some private dependency that includes `activesupport` as a dependency, which shoulda-matchers also depends on
end

group :test do
  gem 'shoulda-matchers'
end
```

**Comment by agrobbin** [comment]
I am going to close this, as I've managed to track down that this is a subtle bug with Bundler itself (rubygems/rubygems#7540) when multiple sources and vendoring of gems are used. Sorry for the confusion!

**Comment by jeffwidman** [comment]
Thanks for circling back to let us know!

You're in good hands with @deivid-rodriguez  :smile:

**Comment by agrobbin** [comment]
@jeffwidman so while this was fixed upstream in Bundler v2.5.9, it seems like Dependabot may not be using that version when doing its `Gemfile.lock` updates? Is it possible to confirm what version of Bundler is being used?

**Comment by agrobbin** [comment]
Closing again, as it looks like this has now been resolved in Dependabot as well!
